### PR TITLE
[SPARK-27563][SQL][TEST] automatically get the latest Spark versions in HiveExternalCatalogVersionsSuite

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

We can get the latest downloadable Spark versions from https://dist.apache.org/repos/dist/release/spark/

## How was this patch tested?

manually.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @srowen @felixcheung @HyukjinKwon @gatorsmile

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala [hunk 1]
```diff
@@ -206,7 +206,16 @@ class HiveExternalCatalogVersionsSuite extends SparkSubmitTestUtils {
...
   // Tests the latest version of every release line.
-  val testingVersions = Seq("2.3.3", "2.4.2")
+  lazy val testingVersions: Seq[String] = {
+    import scala.io.Source
+    val versions = Source.fromURL("https://dist.apache.org/repos/dist/release/spark/").mkString
+      .split("\n")
+      .filter(_.contains("""<li><a href="spark-"""))
+      .map("""<a href="spark-(\d.\d.\d)/">""".r.findFirstMatchIn(_).get.group(1))
+      .filter(_ < org.apache.spark.SPARK_VERSION)
+    logInfo(s"Testing ${org.apache.spark.SPARK_VERSION} with ${versions.mkString(", ")}.")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala]: we can check the testing log to make sure we picked the right versions to test.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala [hunk 2]
```diff
@@ -206,7 +206,16 @@ class HiveExternalCatalogVersionsSuite extends SparkSubmitTestUtils {
...
 object PROCESS_TABLES extends QueryTest with SQLTestUtils {
   // Tests the latest version of every release line.
-  val testingVersions = Seq("2.3.3", "2.4.2")
+  lazy val testingVersions: Seq[String] = {
+    import scala.io.Source
+    val versions = Source.fromURL("https://dist.apache.org/repos/dist/release/spark/").mkString
+      .split("\n")
+      .filter(_.contains("""<li><a href="spark-"""))
+      .map("""<a href="spark-(\d.\d.\d)/">""".r.findFirstMatchIn(_).get.group(1))
+      .filter(_ < org.apache.spark.SPARK_VERSION)
```
Comment (felixcheung) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala]: isn;t this always true?

**Review by felixcheung — APPROVED** [review]

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala]: Nice! @cloud-fan

Comment (HyukjinKwon) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala]: Cool!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala [hunk 3]
```diff
@@ -206,7 +206,16 @@ class HiveExternalCatalogVersionsSuite extends SparkSubmitTestUtils {
 
 object PROCESS_TABLES extends QueryTest with SQLTestUtils {
   // Tests the latest version of every release line.
-  val testingVersions = Seq("2.3.3", "2.4.2")
+  lazy val testingVersions: Seq[String] = {
+    import scala.io.Source
+    val versions = Source.fromURL("https://dist.apache.org/repos/dist/release/spark/").mkString
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala]: BTW, what if this link is temporarily unavailable?
Even if this is a potential issue, I think we should merge this and see if the issue is virtually ignorable/trivial or not.

**Comment by srowen** [comment]
This assumes that dist has only the latest versions in active branches, which it should, but depends on us deleting previous releases diligently. Not sure if it's in the release process. Ideally this would have further logic to only pick the latest version per branch. But not sure if that's complicated to write. Right now this would actually run an extra set of tests.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogVersionsSuite.scala]: This PR should be merged to all the active branches: 2.3, 2.4 and master. Branch 2.3 should not test with Spark 2.4.x

**Comment by cloud-fan** [comment]
@srowen yea we may need to run an extra set of tests for a while during the release. I think it's ok as the failure window is short and the release doesn't happen very frequently.

**Comment by cloud-fan** [comment]
seems `logInfo` doesn't work in test suites. I searched to log and found `SparkContext:54 - Running Spark version 2.3.3` and `SparkContext:54 - Running Spark version 2.4.2`, so it does work.

**Review by srowen — APPROVED** [review]
Looks fine; the 2.4.1 release has been removed from dist (thanks)

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by cloud-fan** [comment]
I've backported it to 2.4/2.3, so that we can completely ignore it during release process. I'll update https://spark.apache.org/release-process.html soon.

**Comment by HyukjinKwon** [comment]
Oops thanks

**Comment by dongjoon-hyun** [comment]
Hi, All.
Unfortunately, `3.0.0-preview` breaks this patch. I made a PR to recover it.
- https://github.com/apache/spark/pull/26428

**Pull request closed without merging**